src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,276 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\InvalidArgumentException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\LockExpiredException;
+use Symfony\Component\Lock\Exception\LockStorageException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\StoreInterface;
+
+/**
+ * MongoDbStore is a StoreInterface implementation using MongoDB as a storage
+ * engine.
+ *
+ * @author Joe Bennett <joe@assimtech.com>
+ */
+class MongoDbStore implements StoreInterface
+{
+    private $mongo;
+    private $options;
+    private $collection;
+
+    /**
+     * @param \MongoDB\Client $mongo
+     * @param array           $options
+     *
+     * database:    The name of the database [required]
+     * collection:  The name of the collection [default: lock]
+     * initialTtl:  The expiration delay of locks in seconds [default: 300.0]
+     *
+     * CAUTION: The locked resouce name is indexed in the _id field of the
+     * lock collection.
+     * An indexed field's value in MongoDB can be a maximum of 1024 bytes in
+     * length inclusive of structural overhead.
+     *
+     * @see https://docs.mongodb.com/manual/reference/limits/#Index-Key-Limit
+     *
+     * CAUTION: This store relies on all client and server nodes to have
+     * synchronized clocks for lock expiry to occur at the correct time.
+     * To ensure locks don't expire prematurely; the lock TTL should be set
+     * with enough extra time to account for any clock drift between nodes.
+     * @see self::createTTLIndex()
+     *
+     * writeConcern, readConcern and readPreference are not specified by
+     * MongoDbStore meaning the collection's settings will take effect.
+     * @see https://docs.mongodb.com/manual/applications/replication/
+     */
+    public function __construct(\MongoDB\Client $mongo, array $options)
+    {
+        if (!isset($options['database'])) {
+            throw new InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+            'initialTtl' => 300.0,
```

Comment:
I really wonder if options for lock should be merge with options for database connection?

See PdoTrait from cache components.

What is the opinion of other contributors?

Reply:
I originally had `initialTtl` as a separate constructor parameter, I brought it in to `$options` when you suggested `drift` should be.

I personally think it belongs outside of `$options`.

See also: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Adapter/PdoAdapter.php